src/Microsoft.OData.Client/ALinq/UriWriter.cs

```diff
@@ -582,14 +585,121 @@ internal void VisitCustomQueryOptions(Dictionary<ConstantExpression, ConstantExp
             }
         }
 
+        /// <summary>
+        /// ApplyQueryOptionExpression visit method.
+        /// </summary>
+        /// <param name="applyQueryOptionExpr">ApplyQueryOptionExpression expression to visit</param>
+        internal void VisitQueryOptionExpression(ApplyQueryOptionExpression applyQueryOptionExpr)
+        {
+            if (applyQueryOptionExpr.Aggregations.Count == 0)
+            {
+                return;
+            }
+
+            // E.g. filter(Amount gt 1)
+            string filterTransformation = ConstructFilterTransformation(applyQueryOptionExpr);
+            // E.g. aggregate(Prop with sum as SumProp, Prop with average as AverageProp)
+            string aggregateTransformation = ConstructAggregateTransformation(applyQueryOptionExpr.Aggregations);
+
+            string applyExpression = string.IsNullOrWhiteSpace(filterTransformation) ? string.Empty : filterTransformation + "/";
+            applyExpression += aggregateTransformation;
+
+            this.AddAsCachedQueryOption(UriHelper.DOLLARSIGN + UriHelper.OPTIONAPPLY, applyExpression);
+        }
+
+        /// <summary>
+        /// Constructs a $apply filter transformation.
+        /// E.g. $apply=filter(Amount gt 1)
+        /// </summary>
+        /// <param name="applyQueryOptionExpr">ApplyQueryOptionExpression expression</param>
+        /// <returns>A filter transformation</returns>
+        private string ConstructFilterTransformation(ApplyQueryOptionExpression applyQueryOptionExpr)
+        {
+            if (applyQueryOptionExpr.PredicateConjuncts.Count == 0)
+            {
+                return string.Empty;
+            }
+
+            return "filter(" + this.ExpressionToString(applyQueryOptionExpr.GetPredicate(), /*inPath*/ false) + ")";
+        }
+
+        /// <summary>
+        /// Constructs a $apply aggregate transformation.
+        /// E.g. $apply=aggregate(Prop with sum as SumProp, Prop with average as AverageProp)
+        /// </summary>
+        /// <param name="aggregations">List of aggregations.</param>
+        /// <returns>The aggregate tranformation.</returns>
+        private string ConstructAggregateTransformation(IList<ApplyQueryOptionExpression.Aggregation> aggregations)
+        {
+            StringBuilder aggregateBuilder = new StringBuilder();
+
+            aggregateBuilder.Append(UriHelper.AGGREGATE);
+            aggregateBuilder.Append(UriHelper.LEFTPAREN);
+            int i = 0;
+
+            while (true)

```

Comment:
> while (true) [](start = 11, length = 13)

why don't use foreach (...) ?

Reply:
@xuzhg This is one scenario where traversing the collection using the indexer worked better. In a single statement, we're able increment the counter and check if its time to exit to avoid adding a trailing comma. The trailing comma would make the Uri invalid. Re:
```csharp
    if (++i == aggregations.Count)
    {
        break;
    }

    aggregateBuilder.Append(UriHelper.COMMA);
```